Enhance UX of Searchbar and make code simpler and more streamlined

tacigar opened the pull request:
resolve #3007

I think the current search bar has the following problems

* Because it uses external libraries such as react-select, the design and keyboard operations are not fully controlled by lens side.
* The implementation is more complex than it needs to be.
* The design is also a bit confusing (especially lookback and limit).
* It is hard to make unit tests because its implementation is too complex.

So I decided to reimplement it from scratch.

This PR will accomplish the following

* Independence from difficult-to-control external libraries.
* Code will be simpler and the code lines will be fewer.
* Become testable.
* UI will probably be easier to understand.
* It will be more extensible & customizable in the future because it does not depend on external libraries.
* TypeScriptize search bar!

---

# GIFs

### Lookback

![lookback](https://user-images.githubusercontent.com/19551419/87039322-f2bad680-c229-11ea-9b77-6f83154619fb.gif)

### Searchbar

![searchbar](https://user-images.githubusercontent.com/19551419/87039350-fe0e0200-c229-11ea-8a76-bf676859df72.gif)

tacigar committed:
Enhance UX of Searchbar and make code simpler and more streamlined

tacigar committed:
Remove unused files

tacigar commented:
I need to make unit tests.

codefromthecrypt commented:
this feels very good. thanks @tacigar!

tacigar committed:
TypeScriptize ExplainBox

tacigar committed:
Remove unused imports

tacigar committed:
Add LookbackMenu's unit tests

tacigar committed:
Add @testing-library/react-hooks

tacigar committed:
Add DiscoverPageContent's unit tests

tacigar committed:
Add SearchBar's unit tests

tacigar committed:
Update CriterionBox

tacigar committed:
Add zIndex props to SuggestionList

jorgheymans commented:
This looks really great, some early feedback to keep your energy flowing :-) If this is too early I apologize, feel free to let me know then and i'll back off for a bit until you mark me as reviewer.

- small bleed-through issue on FF (Ubuntu)
![image](https://user-images.githubusercontent.com/193792/80994935-bf308a00-8e3d-11ea-9fe1-1328f3c4dfc9.png)
- when the first element in the search bar is the `spanName` then there is no auto complete, I guess this is because there are too many spanNames and we can't load them all ? In any case, users are going to be confused
- funny behaviour:
![lens](https://user-images.githubusercontent.com/193792/80996374-fd2ead80-8e3f-11ea-844c-212e4e4794dc.gif)

tacigar committed:
Update

tacigar committed:
Simplify

tacigar committed:
More simple

tacigar commented:
@jorgheymans 
Thank you for your review!

> * small bleed-through issue on FF (Ubuntu)

Fixed!

<img width="493" alt="スクリーンショット 2020-05-05 20 58 14" src="https://user-images.githubusercontent.com/19551419/81063636-2f5d0f80-8f13-11ea-9246-a932e7000527.png">

> * when the first element in the search bar is the spanName then there is no auto complete, I guess this is because there are too many spanNames and we can't load them all ? In any case, users are going to be confused

A service name must be specified to load span name suggestions.
https://zipkin.io/zipkin-api/#/default/get_spans
But as you said, current behavior may be confusing to users.
I have 2 solutions in mind.

* Leave the behavior as it is and display a friendly manual somewhere.
* Disable the selection of span name or remote service name before the service name is selected.

WDYT?

> * funny behaviour:

Fixed the behavior.

![search](https://user-images.githubusercontent.com/19551419/81126360-0326aa80-8f76-11ea-8b02-8a96cf33f450.gif)

tacigar committed:
Fetch autocomplete values when autocomplete key is selected

tacigar committed:
Supports strings containing "="

tacigar commented:
Finally, I will make a unit test of `CriterionBox`.
This will require checking out a variety of scenarios and will take some time.

codefromthecrypt commented:
thanks for the status update!

anuraaga commented on zipkin-lens/src/components/DiscoverPage/DiscoverPage.tsx:
```diff
@@ -0,0 +1,65 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+import React from 'react';
```
For new files I've been trying to alphabetize imports, can you try it?

tacigar replied:
What did you use for alphabetizing?
VSCode extension?

anuraaga replied:
Manual :) Eventually we can extend the eslint to do it automatically though, in other projects it's been automatic for me.

tacigar replied:
I got it!

anuraaga commented on zipkin-lens/src/components/DiscoverPage/DiscoverPageContent.tsx:
```diff
@@ -0,0 +1,386 @@
+interface Props extends RouteComponentProps {}
+
+// Export for testing
+export const useQueryParams = (history: History, location: Location) => {
```
Instead of passing in `history`, `location` you should be able to call `useHistory`, `useLocation` in this hook

anuraaga commented on zipkin-lens/src/components/DiscoverPage/DiscoverPageContent.tsx:
```diff
@@ -0,0 +1,386 @@
+    return ret;
+  }, [location.search]);
+
+  const lookback = useMemo<Lookback | null>(() => {
```
Think it's more conventional to use `undefined` instead of `null`. For example, in interface definitions there's shorthand for it

```ts
interface Foo {
  bar: boolean?;
}
```

anuraaga commented on zipkin-lens/src/components/DiscoverPage/DiscoverPageContent.tsx:
```diff
@@ -0,0 +1,386 @@
+      return null;
+    }
+    const data = fixedLookbackMap[lookback];
+    if (!data) {
```
I think we should also handle when the lookback is milliseconds, not short durations tring

anuraaga replied:
Are we going to handle this one? I think lookback can be a millisecond value if the user specifies the environment variable for it.

tacigar replied:
Ah, I understand what you are saying!
You say we should allow users to specify lookback in milliseconds manually, right?

I agree with you.
What do you think of the following UI?
Users can specify millisecond manually.

![スクリーンショット 2020-07-10 19 12 18](https://user-images.githubusercontent.com/19551419/87143575-4d633980-c2e1-11ea-909e-c7d0e17c777a.png)

anuraaga reviewed (commented):
Did an initial review, overall looks good and nice reduction in lines of code.

One overall point, I don't see handling of the `defaultLookback` UiConfig. From what I understand, we can have a fixed lookback that is not one of the values in our map if the server has that flag set.

jorgheymans commented:
@tacigar 
> Disable the selection of span name or remote service name before the service name is selected.

i would go for this option, it's pretty drastic though and it it's bound to generate the same 'huh?' from users as not autocompleting it. Ideally there is some visual relationship between service name and span name that makes it clear that you cannot filter on the latter without first setting the former.

codefromthecrypt commented:
@tacigar no pressure but probably a release will happen next couple days. lemme know if we should wait for this

tacigar commented:
Sorry for late...
Implementation is still going to take some time...
Please release 🙇

tacigar committed:
Add styled-components package

tacigar committed:
Setup styled-component's ThemeProvider and type definition

tacigar committed:
Use styled-components for SuggestionList

tacigar committed:
Change DataTimePicker size

tacigar committed:
Allow to set up a millisecond lookbacks

tacigar committed:
Handle defaultLookback of UiConfig

tacigar committed:
Fix UiConfig bug

tacigar committed:
Handle 'tags' correctly

tacigar committed:
Update DiscoverPageContent unit tests

tacigar committed:
Alphabetize imports

tacigar committed:
Use styled-components for CriterionBox and fix delete-button's style bug

tacigar committed:
Add animations to CriterionBox

tacigar committed:
Suggestions

tacigar committed:
Fix z-index bug

tacigar committed:
Fix query parameter bug

tacigar committed:
Add HowToUse component

tacigar committed:
Parse duration criteria

tacigar committed:
Update SearchBar unit test

tacigar committed:
Setup queryLimit using config.json

tacigar committed:
Refactor DiscoverPageContent

tacigar committed:
Update tags explanation

tacigar committed:
tags -> tagQuery

tacigar committed:
Resolve conflicts

tacigar committed:
Update copyright

tacigar committed:
Update package-lock.json

tacigar committed:
Add CriterionBox unit tests

tacigar committed:
use useLocation and useHistory

tacigar commented:
PTAL 🙇

codefromthecrypt commented:
is the description including giphy up to date?

tacigar commented:
Sorry, not yet...
And please wait to review 🙇 

I just realized that I forgot to implement this feedback by @jorgheymans ...

https://github.com/openzipkin/zipkin/pull/3076#issuecomment-626000375

tacigar committed:
Show spanName and remoteServiceName after serviceName is selected

tacigar commented:
PTAL 🙇‍♂️

codefromthecrypt commented:
This is really nice! one thing I would like if possible.. after you select service name, and hit enter. I would like it it to execute the search, or allow me to hit enter again to do that. Is this possible? right now I have to move the mouse to click..

codefromthecrypt commented on zipkin-lens/src/components/DiscoverPage/DiscoverPageContent.test.jsx:
```diff
@@ -0,0 +1,215 @@
+      { key: 'maxDuration', value: '100ms' },
+      // AnnotationQuery
+      { key: 'key3', value: 'value3' },
+      { key: 'tagQuery', value: 'key1=value1 and key2' },
```
I know tagQuery doesn't match the old rest api annotationQuery, but chatted with @tacigar and I think these days almost never someone would put an "annotation" in there.

anyone object to having the UI say tagQuery not annotationQuery? (the rest api is the same regardless as not important enough to change) @openzipkin/core

codefromthecrypt reviewed (approved):
I like it. I also tested with auto-complete tags to make sure they still show up.

ex.
```java
AUTOCOMPLETE_KEYS=mvc.controller.class java -jar ./zipkin-server/target/zipkin-server-*slim.jar
```

![Screen Shot 2020-07-10 at 7 45 25 AM](https://user-images.githubusercontent.com/64215/87101317-63d9a880-c281-11ea-8c9d-741c7265bb92.png)
![Screen Shot 2020-07-10 at 7 44 57 AM](https://user-images.githubusercontent.com/64215/87101321-65a36c00-c281-11ea-8325-e48b273d2fe0.png)

tacigar committed:
Search traces when enter key is pressed

tacigar commented:
Sorry, I found a bug... I'll fix it.

tacigar committed:
Add @types/shortid

tacigar committed:
Fix key props bug of SearchBar

tacigar commented:
@adriancole 
You can now fetch with the Enter key :)

![enterkey](https://user-images.githubusercontent.com/19551419/87131318-31ef3300-c2cf-11ea-85f0-6ce80bfb96c6.gif)

codefromthecrypt commented:
heart heart!

anuraaga commented on zipkin-lens/src/components/DiscoverPage/DiscoverPageContent.tsx:
```diff
@@ -0,0 +1,463 @@
+  const [tempLookback, setTempLookback] = useState<Lookback>(
+    lookback || {
+      type: 'fixed',
+      // If defaultLookback in config.json is incorrect, use 15m as an initial value.
```
What's an incorrect default lookback?

anuraaga commented on zipkin-lens/src/components/DiscoverPage/LookbackMenu.tsx:
```diff
@@ -0,0 +1,233 @@
+
+  const handleDialogClose = useCallback(() => {
+    // Use setTimeout to change isOpeningDialog state after
+    // handleOutsideClick callback function is executed.
```
Why is this important?

tacigar replied:
If there is no setTimeout, LookbackMenu will also be closed when the DateTimePicker's dialog is clicked.

![without-settimeout](https://user-images.githubusercontent.com/19551419/87141773-89e16600-c2de-11ea-8728-096b64d76ac0.gif)

I use setTimeout to achieve the following behavior

![with-settimeout](https://user-images.githubusercontent.com/19551419/87141863-b0070600-c2de-11ea-92f8-5a5653b5e9bd.gif)

anuraaga commented on zipkin-lens/src/components/DiscoverPage/LookbackMenu.tsx:
```diff
@@ -0,0 +1,233 @@
+  const handleOutsideClick = useCallback(
+    (event: any) => {
+      if (!isOpeningDialog) {
+        if (!el.current) {
```
Looks like you can combine the ifs

anuraaga commented on zipkin-lens/src/components/DiscoverPage/SearchBar/CriterionBox.tsx:
```diff
@@ -0,0 +1,458 @@
+        onDelete();
+        return;
+      }
+      let ss = fixedText.split('=');
```
Can we use a more descriptive name than `ss`? Not sure what `ss` stands for here

anuraaga commented on zipkin-lens/src/components/DiscoverPage/SearchBar/SearchBar.tsx:
```diff
@@ -0,0 +1,262 @@
+  // If the value is -1, there is no criterion being edited.
+  const [criterionIndex, setCriterionIndex] = useState(-1);
+
+  const handleCriterionFocus = (index: number) => () => {
```
As far as I can tell this callback pattern has bad performance since the functions all get initialized every render. Instead of using the `index` factory can't you do something like `useCallback((criterion) => criteria.find(c => c.id == criterion.id)`?

tacigar committed:
Combine ifs

jcchavezs commented:
I see the custom loopback supporting millisecond (in the input), maybe that
is too much as seconds is just fine?

fre. 10. jul. 2020, 11:55 skrev tacigar <notifications@github.com>:

> *@tacigar* commented on this pull request.
> ------------------------------
>
> In zipkin-lens/src/components/DiscoverPage/LookbackMenu.tsx
> <https://github.com/openzipkin/zipkin/pull/3076#discussion_r452745366>:
>
> > +  lookback,
> +}) => {
> +  const classes = useStyles();
> +
> +  // LookbackMenu is closed when click on the outside of the component.
> +  // This state is needed to prevent LookbackMenu component from closing
> +  // when the dialog of DateTimePicker is clicked.
> +  const [isOpeningDialog, setIsOpeningDialog] = useState(false);
> +
> +  const handleDialogOpen = useCallback(() => {
> +    setIsOpeningDialog(true);
> +  }, []);
> +
> +  const handleDialogClose = useCallback(() => {
> +    // Use setTimeout to change isOpeningDialog state after
> +    // handleOutsideClick callback function is executed.
>
> If there is no setTimeout, LookbackMenu will also be closed when the
> DateTimePicker's dialog is clicked.
>
> [image: without-settimeout]
> <https://user-images.githubusercontent.com/19551419/87141773-89e16600-c2de-11ea-8728-096b64d76ac0.gif>
>
> I use setTimeout to achieve the following behavior
>
> [image: with-settimeout]
> <https://user-images.githubusercontent.com/19551419/87141863-b0070600-c2de-11ea-92f8-5a5653b5e9bd.gif>
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin/pull/3076#discussion_r452745366>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXOYAVHKY5WMJSNRQOOHXTR23QPVANCNFSM4MYG2Y6A>
> .
>

tacigar commented:
I see... 👍

tacigar committed:
Don't support milliseconds in custom lookback

tacigar committed:
Update unit tests

jorgheymans commented:
thanks @tacigar ! Some feedback :

- when manually clearing spanName , it should reappear in the list:
![no-tagname-anymore](https://user-images.githubusercontent.com/193792/87220085-5267cd80-c361-11ea-830e-6a3715f4fe9e.gif)

- the empty criteria box seems an invalid state, but i'm not sure how to reproduce it consistently:
![zipkin-invalid-state](https://user-images.githubusercontent.com/193792/87220100-79260400-c361-11ea-80b3-5b5dabb55756.png)

- if you don't select a value for service name and click outside of the widget it stays like this. Not sure if intended ?
![zipkin-service-name](https://user-images.githubusercontent.com/193792/87220106-893de380-c361-11ea-9f76-d9c5dd1d8da0.png)

- I noticed some suspicious behaviour pressing reload in the browser, and using 'back' button to go from trace screen back to search overview it may be in play when state gets corrupted not sure.

- Should the search button go next to the 'limit' field to reduce mouse travel ? Or make it more clear somehow that ctrl-enter does the same thing.

Overall good improvements !! I'll try to play with it more later today.

tacigar commented:
Thank you for feedback! @jorgheymans 

> * when manually clearing spanName , it should reappear in the list:

Sorry. I couldn't reproduce it.
Maybe it has been fixed in the latest commit 🤔 

> * the empty criteria box seems an invalid state, but i'm not sure how to reproduce it consistently:

The criteria box is synchronized with the query parameter.
If the query parameter has an empty criterion, an empty criteria box will be displayed.
For example, if the query parameter is `?serviceName=serviceA&=&spanName=spanA`, then the criteria boxes will be as follows:

![スクリーンショット 2020-07-13 13 41 26](https://user-images.githubusercontent.com/19551419/87271786-43268280-c50f-11ea-9d94-180fe196fbff.png)

I could not reproduce generating this empty criterion from the UI by normal way.
Maybe this has been fixed in the latest commit...

> * if you don't select a value for service name and click outside of the widget it stays like this. Not sure if intended ?

About autocompleteKeys, it is an intended behavior.
This is because autocompleteKeys will eventually be converted to annotationQuery, so criterion without value are sufficient.

But for all other keys(serviceName, spanName, ...), value is needed.
So I'll fix it.

> * I noticed some suspicious behaviour pressing reload in the browser, and using 'back' button to go from trace screen back to search overview it may be in play when state gets corrupted not sure.

hmmm...I could not reproduce it... 🤔 

> * Should the search button go next to the 'limit' field to reduce mouse travel ? Or make it more clear somehow that ctrl-enter does the same thing.

Now, if you press enter when the criteria box is not focused, searching will start 👍

anuraaga commented:
Supporting custom lookback in the form seems nice, but I'm more concerned with the default lookback parameter, like here

https://github.com/openzipkin/zipkin/blob/master/zipkin-server/src/main/java/zipkin2/server/internal/ui/ZipkinUiConfiguration.java#L141

There was code specifically dealing with a milliseconds lookback that could come from that flag before but I can't find it in the new version, can you point me at it? AFAIK that flag can be any millliseconds, not our fixed lookbacks.

anuraaga commented on zipkin-lens/src/components/DiscoverPage/DiscoverPageContent.test.jsx:
```diff
@@ -0,0 +1,221 @@
+  it('should initialize the lookback using config.json', () => {
+    const { getAllByText } = render(<DiscoverPageContent />, {
+      uiConfig: {
+        defaultLookback: 60 * 1000 * 5, // 5m
```
More specifically, we should have a test where this is set to an arbitrary milliseconds like 1234567

tacigar commented:
I got it 🙇

mrajah-twttr commented:
this looks very cool.thanks for taking the time to redo this. 

i got a chance to play around with the new experience and noticed a couple of issues
1. if you search by service name and then you try to edit the box and remove all the content (quick way to clear the service name you entered) the "servicename" option is not present any more in suggested options (i see spanName and other options which are assuming service name is present)? does it make sense to keep search criteria and search text in separate text boxes instead of a single text box with servicename=<abcdservice>? i think this is the same issue jorgheymans noted
2. does it make sense to not have the user click on the + on load and open up with  the  the first search box ready to be filled.
3. i believe this is an issue even in existing UI but when you start typing the service name the characters dont show up right away until the lookahead is returned making it feel like a hang at times. since we are in the same code base would be great if we can fix it?

overall i see this as a nice way forward, excited to push it to our customers when it lands.

jorgheymans commented:
Some more feedback:

- seems the last set dropdown is auto expanded when pressing back from the trace detail view:
![criteria-auto-expanding](https://user-images.githubusercontent.com/193792/88456400-d5b71200-ce7d-11ea-8b0b-ef583acae9a8.gif)

- clicking on the unit in the duration makes the field disappear, could not reproduce consistently but happened often:
![duration-disappearing](https://user-images.githubusercontent.com/193792/88456418-fed7a280-ce7d-11ea-92a0-2bc6ea1f4284.gif)

tacigar committed:
Support specifying a lookback in any millisecond

tacigar commented:
> if you search by service name and then you try to edit the box and remove all the content (quick way to clear the service name you entered) the "servicename" option is not present any more in suggested options (i see spanName and other options which are assuming service name is present)? does it make sense to keep search criteria and search text in separate text boxes instead of a single text box with servicename=? i think this is the same issue jorgheymans noted

I see! I understand!

> does it make sense to not have the user click on the + on load and open up with the the first search box ready to be filled.
i believe this is an issue even in existing UI but when you start typing the service name the characters dont show up right away until the lookahead is returned making it feel like a hang at times. since we are in the same code base would be great if we can fix it?

I see...
You mean the searchbar does not have to be shown until serviceNames has been loaded, right?
It's OK to me.
But the same problem occurs with `spanNames` and `remoteServiceNames`.
The list of spanName and remoteServiceName candidates are loaded after the serviceName associated with them has been determined.
So there is always a waiting time for the list of spanName and remoteServiceName to load when we input criteria to searchbar.
So for now, I'm not going to handle that issue in this PR.
We can change it later.

tacigar committed:
Fix bug

tacigar committed:
Fix DiscoverPageContent unit test

tacigar commented:
> seems the last set dropdown is auto expanded when pressing back from the trace detail view:

Yeah...I removed this feature.
I thought it wasn't necessary :)

> clicking on the unit in the duration makes the field disappear, could not reproduce consistently but happened often:

Ah, I see...
I removed the unit's dropdown feature because keyboard is faster and easier than dropdown 😢

tacigar committed:
Add defaultLookback test

tacigar committed:
Fix CriterionBox behavior

tacigar committed:
Merge branch 'master' into tacigar/search-bar

tacigar committed:
Optimize

tacigar commented:
PTAL 🙇

tacigar committed:
Change design a bit

jorgheymans commented:
Thanks @tacigar ! Some more comments

- when doing browser-back button from the trace detail screen, an empty search criteria box always appears. 
![Peek 2020-08-06 15-25](https://user-images.githubusercontent.com/193792/89537434-a39c9d00-d7f9-11ea-9749-ae0076061fae.gif) 
- When editing for example the span name, should we not again display the full list straight away instead of letting users do backspace to erase and then start matching against the list entries ? Just a thought, hard to explain let me know if it does not make sense.
![Peek 2020-08-06 15-53](https://user-images.githubusercontent.com/193792/89540361-35f27000-d7fd-11ea-8dd1-31ada2143cc8.gif)

tacigar commented:
> when doing browser-back button from the trace detail screen, an empty search criteria box always appears.

Curious... 🤔 
I could not reproduce that behavior.
Could you tell me the URL when the empty criteria are displayed?

> When editing for example the span name, should we not again display the full list straight away instead of letting users do backspace to erase and then start matching against the list entries ? Just a thought, hard to explain let me know if it does not make sense.

I understand what you mean and I agree with you!
But implementation is going to be more complicated than it looks... 😢 
Is it possible to make it future work?

jorgheymans commented:
> > when doing browser-back button from the trace detail screen, an empty search criteria box always appears.
> 
> Curious... thinking
> I could not reproduce that behavior.
> Could you tell me the URL when the empty criteria are displayed?
Here is how to reproduce
* start off at http://localhost:9411/zipkin/
* add servicename, then span name, then press enter at this moment:
![image](https://user-images.githubusercontent.com/193792/89577949-b763f580-d831-11ea-968c-76bd15a955ee.png)
* then click on one of the traces
* then browser-back
* the URL is now http://localhost:9411/zipkin/?serviceName=zipkin-server&spanName=get+%2Fzipkin%2Fapi%2Fv2%2Fremoteservices&=&lookback=15m&endTs=1596744750320&limit=10 
 
> I understand what you mean and I agree with you!
> But implementation is going to be more complicated than it looks... cry
> Is it possible to make it future work?
:+1:

tacigar commented:
@jorgheymans Thank you!
I could reproduce that behavior.
I'll fix it.

tacigar committed:
Fix bugs

tacigar commented:
@jorgheymans fixed 🙇

tacigar commented:
Oops! test failed

codefromthecrypt commented:
@jorgheymans merge if you think this addresses your comments, ya?

tacigar commented:
Sorry, please wait for merge.
I'll update unit tests which are failing soon.

tacigar committed:
Update unit tests

tacigar commented:
updated 🙇

tacigar commented:
Lens' unit tests succeeded.
![スクリーンショット 2020-08-07 21 08 02](https://user-images.githubusercontent.com/19551419/89644136-41739300-d8f2-11ea-9088-d3d574587bc0.png)

codefromthecrypt commented:
very cool. we will make the next release a minor so we can announce the smoother stuff
